File: src/lcgc.php
<?php

define('BUFFER_SIZE', 1024);

require_once(dirname(__FILE__).'/../lib/lcg_cipher.php');

function usage($cmd, $err = '') {
	if(strlen($err) > 0) {
		fprintf(STDERR, "Error: %s\n", $err);
	}
	fprintf(STDERR, "Usages : %s cipher|decipher <password> [input file [output file]]\n", $cmd);
	fprintf(STDERR, "         %s prng <password> <count>\n", $cmd);
	fprintf(STDERR, "         %s debug <password>\n", $cmd);
	fprintf(STDERR, "         %s cycles <password>\n", $cmd);
	exit(1);
}

function do_cipher_decipher($argc, $argv) {
	$password = $argv[2];
	$mode = $argv[1];
	$in_fd = false;
	$out_fd = false;
	
	if($argc >= 4) {
		if($argv[3] != '-') {
			$in_fd = fopen($argv[3], 'rb');
			if($in_fd === false)
				usage($argv[0], 'Cannot open '.$argv[3].' for reading!');
		}
	}
	if($argc == 5) {
		if($argv[4] != '-') {
			$out_fd = fopen($argv[4], 'wb');
			if($out_fd === false)
				usage($argv[0], 'Cannot open '.$argv[4].' for writing!');
		}
	}
	
	if($in_fd === false)
		$in_fd = STDIN;
	if($out_fd === false)
		$out_fd = STDOUT;
	
	$lc = new LCGCipher($password);
	
	if($mode == 'cipher') {
		$data = fread($in_fd, BUFFER_SIZE);
		while($data !== false) {
			$cdata = $lc->cipher($data);
			fwrite($out_fd, $cdata);
			if(feof($in_fd))
				break;
			$data = fread($in_fd, BUFFER_SIZE);
		}
	}
	else if($mode == 'decipher') {
		$data = fread($in_fd, BUFFER_SIZE);
		while($data !== false) {
			$ddata = $lc->decipher($data);
			fwrite($out_fd, $ddata);
			if(feof($in_fd))
				break;
			$data = fread($in_fd, BUFFER_SIZE);
		}
	}
	
	if($in_fd !== STDIN)
		fclose($in_fd);
	if($out_fd !== STDOUT)
		fclose($out_fd);
}

function do_prng($argc, $argv) {
	$password = $argv[2];
	$count = $argv[3];
	
	$lc = new LCGCipher($password);
	
	for($i = 0 ; $i < $count ; $i++)
		echo $lc->genRand()."\n";
}

function do_debug($argc, $argv) {
	$password = $argv[2];
	
	$lc = new LCGCipher($password);
	
	echo "a           x           c           m\n";
	
	for($i = 0, $l = count($lc->generators) ; $i < $l ; $i++) {
		echo str_pad($lc->generators[$i]->a, 11).' '.
		     str_pad($lc->generators[$i]->x, 11).' '.
		     str_pad($lc->generators[$i]->c, 11).' '.
		     str_pad($lc->generators[$i]->m, 11)."\n";
	}
}

function math_gcd($a, $b) {
	if($a < $b)
		list($a, $b) = array($b, $a);
	
	while($b != 0) {
		list($a, $b) = array($b, bcmod($a, $b));
	}
	
	return($a);
}

function do_cycles($argc, $argv) {
	$password = $argv[2];
	
	$lc = new LCGCipher($password);
	
	echo "Starting";
	
	$cycles = array();
	for($i = 0, $l = count($lc->generators) ; $i < $l ; $i++) {
		$first = $lc->generators[$i]->gen();
		$prev = 0;
		for($j = 1 ; $j <= $lc->generators[$i]->m ; $j++) {
			$prev = $lc->generators[$i]->gen();
			if($first == $prev)
				break;
		}
		
		if($first != $prev) {
			$first = $lc->generators[$i]->gen();
			for($j = 1 ; $j <= $lc->generators[$i]->m ; $j++) {
				$prev = $lc->generators[$i]->gen();
				if($first == $prev)
					break;
			}
		}
		$j++;
		$cycles[] = ''.$j;
		echo '.';
		fflush(STDOUT);
	}
	
	echo "\n";
	
	$glob_gcd = 1;
	for($i = 0 ; $i < count($cycles) - 1 ; $i++) {
		for($j = $i + 1 ; $j < count($cycles) ; $j++) {
			$glob_gcd = bcmul($glob_gcd, math_gcd($cycles[$i], $cycles[$j]));
		}
	}
	
	echo "GCD(".implode(", ", $cycles).") = $glob_gcd\n";
	
	$cycle = '1';
	foreach($cycles as $c)
		$cycle = bcmul($cycle, $c);
	
	$cycle = bcdiv($cycle, $glob_gcd);
	
	echo "Cycle: $cycle\n";
}

if($argc < 2) {
	usage($argv[0]);
}

$mode = $argv[1];
if($mode == 'cipher' || $mode == 'decipher') {
	if($argc < 3 || $argc > 5)
		usage($argv[0]);
	
	do_cipher_decipher($argc, $argv);
}
else if($mode == 'prng') {
	if($argc != 4)
		usage($argv[0]);
	
	do_prng($argc, $argv);
}
else if($mode == 'debug') {
	if($argc != 3)
		usage($argv[0]);
	
	do_debug($argc, $argv);
}
else if($mode == 'cycles') {
	if($argc != 3)
		usage($argv[0]);
	
	do_cycles($argc, $argv);
}
else {
	usage($argv[0]);
}

File: src/lcgc.cpp
#include "../lib/lcg_cipher.hpp"

#include <cstdio>
#include <cstring>

#define BUFFER_SIZE 1024

void usage(const char *self, const std::string &error = std::string()) {
	if(error.size() > 0)
		printf("Error: %s\n", error.c_str());
	fprintf(stderr, "Usages : %s cipher|decipher <password> [input file [output file]]\n", self);
	fprintf(stderr, "         %s prng <password> <count>\n", self);
	fprintf(stderr, "         %s debug <password>\n", self);
	exit(error.size() == 0 ? 0 : 1);
}

void do_cipher_decipher(int argc, char **argv) {
	LCGCipher *lc;
	uint8_t *rbuffer;
	size_t rbuffer_len;
	std::vector<uint8_t> wbuffer;
	FILE *in_fp = NULL;
	FILE *out_fp = NULL;
	std::string password;
	
	password = argv[2];
	
	if(argc >= 4) {
		if(strcmp(argv[3], "-") != 0) {
			in_fp = fopen(argv[3], "rb");
			if(in_fp == NULL)
				usage(argv[0], std::string("Cannot open ") + argv[3] + " for reading!");
		}
	}
	if(argc == 5) {
		if(strcmp(argv[4], "-") != 0) {
			out_fp = fopen(argv[4], "wb");
			if(out_fp == NULL)
				usage(argv[0], std::string("Cannot open ") + argv[4] + " for writing!");
			rewind(out_fp);
		}
	}
	
	if(in_fp == NULL)
		in_fp = stdin;
	if(out_fp == NULL)
		out_fp = stdout;
	
	try {
		lc = new LCGCipher(password.c_str(), password.size());
	}
	catch(std::string &e) {
		printf("LCGCipher error : '%s'\n", e.c_str());
		exit(1);
	}
	rbuffer = new uint8_t[BUFFER_SIZE];
	
	if(strcmp(argv[1], "cipher") == 0) {
		while((rbuffer_len = fread(rbuffer, sizeof(rbuffer[0]), BUFFER_SIZE, in_fp)) > 0) {
			lc->cipher(rbuffer, rbuffer_len, wbuffer);
			fwrite(wbuffer.data(), sizeof(rbuffer[0]), wbuffer.size(), out_fp);
			wbuffer.clear();
		}
	}
	else {
		while((rbuffer_len = fread(rbuffer, sizeof(rbuffer[0]), BUFFER_SIZE, in_fp)) > 0) {
			lc->decipher(rbuffer, rbuffer_len, wbuffer);
			fwrite(wbuffer.data(), sizeof(rbuffer[0]), wbuffer.size(), out_fp);
			wbuffer.clear();
		}
	}
	
	delete lc;
	delete[] rbuffer;
	if(in_fp != stdin)
		fclose(in_fp);
	if(out_fp != stdout)
		fclose(out_fp);
}

void do_prng(int argc, char **argv) {
	LCGCipher *lc;
	std::string password;
	uint64_t count;
	
	password = argv[2];
	
	try {
		lc = new LCGCipher(password.c_str(), password.size());
	}
	catch(std::string &e) {
		printf("LCGCipher error : '%s'\n", e.c_str());
		exit(1);
	}
	
	sscanf(argv[3], "%llu", &count);
	
	for(uint64_t i = 0 ; i < count ; i++) {
		printf("%llu\n", lc->genRand());
	}
	
	delete lc;
}

void do_debug(int argc, char **argv) {
	LCGCipher *lc;
	std::string password;
	
	password = argv[2];
	
	try {
		lc = new LCGCipher(password.c_str(), password.size());
	}
	catch(std::string &e) {
		printf("LCGCipher error : '%s'\n", e.c_str());
		exit(1);
	}
	
	printf("a         x         c         m\n");
	
	for(int i = 0, l = lc->generators.size() ; i < l ; i++) {
		printf(
			"%-9llu %-9llu %-9llu %-9llu\n",
			lc->generators[i].a,
			lc->generators[i].x,
			lc->generators[i].c,
			lc->generators[i].m
		);
	}
	
	delete lc;
}

int main(int argc, char **argv) {
	if(argc < 2)
		usage(argv[0]);
	
	if(strcmp(argv[1], "cipher") == 0 || strcmp(argv[1], "decipher") == 0) {
		if(argc < 3 || argc > 5)
			usage(argv[0]);
		
		do_cipher_decipher(argc, argv);
	}
	else if(strcmp(argv[1], "prng") == 0) {
		if(argc != 4)
			usage(argv[0]);
		
		do_prng(argc, argv);
	}
	else if(strcmp(argv[1], "debug") == 0) {
		if(argc != 3)
			usage(argv[0]);
		
		do_debug(argc, argv);
	}
	else {
		usage(argv[0], std::string("Unknown command ") + argv[1]);
	}
	
	return(0);
}
